File: templates/typescript-stdio/src/tools.ts
/**
 * Pure, side-effect-free logic for your tools lives here so it stays easy to
 * unit-test. The MCP wiring in `index.ts` just calls these functions.
 */

export interface TextStats {
  characters: number;
  charactersNoSpaces: number;
  words: number;
  lines: number;
  sentences: number;
}

export function textStats(text: string): TextStats {
  const words = text.trim().length === 0 ? [] : text.trim().split(/\s+/);
  const sentences = text.split(/[.!?]+/).filter((s) => s.trim().length > 0);
  return {
    characters: text.length,
    charactersNoSpaces: text.replace(/\s/g, "").length,
    words: words.length,
    lines: text.split(/\r\n|\r|\n/).length,
    sentences: sentences.length,
  };
}

export function ping(): { message: string; timestamp: string } {
  return { message: "pong", timestamp: new Date().toISOString() };
}

export function greeting(name: string): string {
  const clean = name.trim() || "world";
  return `Hello, ${clean}! This greeting was served by an MCP resource.`;
}

File: templates/python-stdio/tools.py
"""Pure, side-effect-free logic for your tools.

Keeping it here (separate from the MCP wiring in ``server.py``) makes it
trivial to unit-test with pytest.
"""

from __future__ import annotations

import re
from datetime import datetime, timezone


def text_stats(text: str) -> dict[str, int]:
    """Count characters, words, lines and sentences in ``text``."""
    words = text.split()
    sentences = [s for s in re.split(r"[.!?]+", text) if s.strip()]
    return {
        "characters": len(text),
        "characters_no_spaces": len(re.sub(r"\s", "", text)),
        "words": len(words),
        "lines": len(text.splitlines()) or 1,
        "sentences": len(sentences),
    }


def ping() -> dict[str, str]:
    """Health check payload."""
    return {
        "message": "pong",
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def greeting(name: str) -> str:
    """Build a personalized greeting."""
    clean = name.strip() or "world"
    return f"Hello, {clean}! This greeting was served by an MCP resource."
